contracts/TradetrustERC721.sol

```diff
@@ -1,72 +1,172 @@
+    require(ownerOf(tokenId) == address(this), "TokenRegistry: Token has not been surrendered");
+    _burnToken(tokenId);
   }
 
-  function sendToken(address to, uint256 _tokenId) public onlyMinter {
-    require(ownerOf(_tokenId) == address(this), "Cannot send token: Token not owned by token registry");
-    _safeTransferFrom(ownerOf(_tokenId), to, _tokenId, "");
+  // TODO: modify this to mint a new token if it doesn't exist, and send a token if it is owned by address(this)
+  // rationale for this is that we currently also have a need for a method that performs the action for
+  // minting to a new title escrow directly, which has a large overlap with this function
+  // make sure to write tests for this and check for access controls
+  // Question: Should we combine the 2 actions into a single function or split them into their own functions instead?
+  function sendToNewTitleEscrow(
+    address beneficiary,
+    address holder,
+    uint256 _tokenId
+  ) public onlyMinter {
+    if (_exists(_tokenId)) {
+      require(ownerOf(_tokenId) == address(this), "TokenRegistry: Token is not owned by registry");
+
+      address newTitleEscrow = _deployNewTitleEscrow(address(this), beneficiary, holder);
+      this.safeTransferFrom(address(this), newTitleEscrow, _tokenId, "");
+    } else {
+      address newTitleEscrow = _deployNewTitleEscrow(address(this), beneficiary, holder);
+      _safeMint(newTitleEscrow, _tokenId);
+    }
   }
 
-}
+  function restoreTitle(
+    address beneficiary,
+    address holder,
+    uint256 tokenId
+  ) external onlyMinter returns (address) {
+    require(_exists(tokenId), "TokenRegistry: Token does not exist for transfer");
+    require(ownerOf(tokenId) == address(this), "TokenRegistry: Token is not owned by registry");
 
-contract CalculateTradeTrustERC721Selector {
-  function calculateSelector() public pure returns (bytes4) {
-    TradeTrustERC721 i;
-    return
-      i.onERC721Received.selector ^
-      i.destroyToken.selector ^
-      i.sendToNewTitleEscrow.selector ^
-      i.sendToken.selector;
+    address newTitleEscrow = _deployNewTitleEscrow(address(this), beneficiary, holder);
+    _registrySafeTransformFrom(address(this), newTitleEscrow, tokenId);
+
+    return newTitleEscrow;
+  }
+
+  function mintTitle(
+    address beneficiary,
+    address holder,
+    uint256 tokenId
+  ) external onlyMinter returns (address) {
+    require(!_exists(tokenId), "Cannot mint title: Token already exists");
+
+    address newTitleEscrow = _deployNewTitleEscrow(address(this), beneficiary, holder);
+    _safeMint(newTitleEscrow, tokenId);
+
+    return newTitleEscrow;
   }
-}
\ No newline at end of file
+
+  function transferTitle(
+    address to,
+    uint256 _tokenId
+  ) public onlyMinter {
+    require(ownerOf(_tokenId) == address(this), "TokenRegistry: Token not owned by token registry");
+    _registrySafeTransformFrom(address(this), to, _tokenId);
+  }
```

Comment:
This `transferTitle()` is originally called `sendToken()` before the renaming.

And so if we split the `sendToNewTitleEscrow()` into 2 separate functions, do we still need this `transferTitle()` (which was `sendToken()`) since we have `restoreTitle` already? I think we can remove this function because it looks like `sendToken()` is used for restoring of titles only, right?

Reply:
https://github.com/Open-Attestation/token-registry/pull/32/commits/9e5f084c4053edbb3a2cd7504db32d22bbe0e54d